common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -318,9 +318,46 @@ private static Collation fetchCollation(int collationId) {
         }
       }
 
-      protected static SparkException collationInvalidNameException(String collationName) {
+      protected static SparkException collationInvalidNameException(
+          String collationName,
+          String[] ICULocaleNames,
+          List<String> validModifiers) {
+        Map<String, String> params = new HashMap<>();
+        params.put("collationName", collationName);
+
+        final int MODIFIER_LENGTH = 3;
+        assert(validModifiers.stream().allMatch(m -> m.length() == MODIFIER_LENGTH));
+
+        // Split modifiers and locale name.
+        String localeName = collationName.toUpperCase();
+        List<String> modifiers = new ArrayList<>();
+        while (validModifiers.stream().anyMatch(localeName::endsWith)) {
+          modifiers.add(localeName.substring(localeName.length() - MODIFIER_LENGTH));
+          localeName = localeName.substring(0, localeName.length() - MODIFIER_LENGTH);
+        }
+
+        // Suggest version with unique modifiers.
+        Collections.reverse(modifiers);
+        modifiers = modifiers.stream().distinct().toList();
+
+        // Remove conflicting settings.
+        if (modifiers.contains("_CI") && modifiers.contains(("_CS"))) {
+          modifiers = modifiers.stream().filter(m -> !m.equals("_CI")).toList();
+        }
+
+        if (modifiers.contains("_AI") && modifiers.contains(("_AS"))) {
+          modifiers = modifiers.stream().filter(m -> !m.equals("_AI")).toList();
+        }
+
+        // Find the closest locale name.
+        final String finalLocaleName = localeName;
+        String closestLocale = Collections.min(List.of(ICULocaleNames), Comparator.comparingInt(
+          c -> UTF8String.fromString(c.toUpperCase()).levenshteinDistance(
+            UTF8String.fromString(finalLocaleName))));
+        params.put("proposal", closestLocale + String.join("", modifiers));
+
```

Comment:
I suggest we add an assertion here to check if the produced proposal is a valid collation.